### PR TITLE
Implement reuse past and present tensor in Attention Ops.

### PR DESCRIPTION
**Pull request by zhanghuanrong**
Implement reuse kv_cache past and present tensor in Attention Ops. 
Unit test for abover feature.
Utilize the reuse kv_cache for past and present tensor in Greedy Search.
Correctness test for it.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 1]
```diff
@@ -168,6 +168,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Hidden dimension of Q, K, V: hidden_size, hidden_size and v_hidden_size",
               AttributeProto::INTS,
               OPTIONAL_VALUE)
+        .Attr("kv_cache_past_present",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:171]: Maybe name it as reuse_cache_buffer, which means the buffer for the past input is reused for the present output.

BTW, we may not need this attribute. Instead, we rely on whether past_sequence_length input exists for such flag.

File: onnxruntime/contrib_ops/cpu/bert/attention_base.h [hunk 2]
```diff
@@ -63,13 +66,15 @@ class AttentionBase {
...
-                     void* parameters) const;
+                     void* parameters,
+                     const Tensor* past_seq_len = nullptr) const;
 
   int num_heads_;                          // number of attention heads
   bool is_unidirectional_;                 // whether every token can only attend to previous tokens.
   std::vector<int64_t> qkv_hidden_sizes_;  // Q, K, V hidden sizes parsed from the qkv_hidden_sizes attribute.
   bool require_same_hidden_size_;          // whether the implementation supports different hidden sizes of Q/K/V.
   bool require_weights_;                   // whether the implementation requires weights for Q/K/V.
+  int kv_cache_past_present_;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention_base.h:77]: use bool type?

File: onnxruntime/contrib_ops/cuda/bert/attention.cc [hunk 3]
```diff
@@ -23,7 +24,9 @@ namespace cuda {
       T,                                                          \
       kCudaExecutionProvider,                                     \
       (*KernelDefBuilder::Create())                               \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())  \
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention.cc:27]: Shall we add .MayInplace(4, 1)?

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 4]
```diff
@@ -201,19 +289,40 @@ Status QkvToContext(
...
+      v = data.present + batches * present_size_per_batch_k;
+    }
 
-  if (nullptr != data.present) {
-    ORT_RETURN_IF_ERROR(
-        LaunchConcatPastToPresent(stream, total_sequence_length, sequence_length, batch_size, qk_head_size, num_heads,
-                                  max_threads_per_block, data.past, k, data.present));
+  } else {
+    ORT_ENFORCE(qk_head_size == v_head_size);
+    if (data.present != data.past) {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:314]: remove later?

File: onnxruntime/contrib_ops/cpu/bert/attention_base.h (hunk 2 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/bert/attention_base.h:77]: And add a comment to explain it

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 5]
```diff
@@ -87,14 +87,100 @@ size_t GetAttentionWorkspaceSize(
...
+                                           const int head_size,
+                                           const int num_heads,
+                                           const int max_threads_per_block,
+                                           const T* biases,
+                                           const T* qkv_buffer,
+                                           T* present) {
+  const dim3 grid(sequence_length, batch_size, 2);  // 2 for k and v
+  int threads = 1 << static_cast<int>(std::ceil(std::log2(head_size)));
+  threads = std::min(threads, max_threads_per_block);
+  const dim3 block(threads, num_heads, 1);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:144]: >threads, num_heads, 1 [](http://example.com/codeflow?start=19&length=21)

threads * num_heads * 1 can be much larger than max_threads_per_block

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 6]
```diff
@@ -87,14 +87,100 @@ size_t GetAttentionWorkspaceSize(
...
+  const int b = blockIdx.y;
+  const int N = blockDim.y;
+  const int S = gridDim.x;
+  const int B = gridDim.y;
+
+  constexpr int M = 3; // Matrix count in qkv
+  const int m = blockIdx.z + 1;  // k = 1, v = 2
+
+  const int NH = N * head_size;
+  const int NHS = NH * S;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:108]: will nvcc optimize this computation

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 1 above)
Comment (zhanghuanrong) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:171]: Nice suggestion.

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu (hunk 6 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:108]: fixed

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 1 above)
Comment (zhanghuanrong) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:171]: Will still use this variable, just change name, the past_sequence_length may better only control its own data for future expansion.

File: onnxruntime/contrib_ops/cpu/bert/attention_base.h (hunk 2 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/bert/attention_base.h:77]: done

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu (hunk 4 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:314]: After found easy testing code bind output, I will remove these.

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu (hunk 5 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:144]: nice find

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu (hunk 6 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:108]: typo above. not sure, yet no hurt here.

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 7]
```diff
@@ -228,6 +228,14 @@ def parse_arguments(argv: Optional[List[str]] = None) -> argparse.Namespace:
     )
     model_group.set_defaults(vocab_mask=False)
 
+    model_group.add_argument(
+        "--past_present_share_buffer",
+        required=False,
+        action="store_true",
+        help="Use kv_cache for past and present, currently work for gpt2 greedy search.",
+    )
+    model_group.set_defaults(past_present_share_buffer=False)
```
Comment (yufenglee) [inline onnxruntime/python/tools/transformers/convert_generation.py:237]: Why not enable it by default?

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 8]
```diff
@@ -126,30 +212,32 @@ Status QkvToContext(
     if (data.bias == nullptr) {
       // gemm_buffer should be BxSx3xNxH => qkv: 3xBxNxSxH
       ORT_ENFORCE(qk_head_size == v_head_size);
+      ORT_ENFORCE(kv_cache_past_present == 0);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu]: >ORT_ENFORCE(kv_cache_past_present == 0); [](http://example.com/codeflow?start=6&length=40)

why kv_cache_past_present has to be 0 if bias i nullptr?

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 7 above)
Comment (zhanghuanrong) [reply onnxruntime/python/tools/transformers/convert_generation.py:237]: First to backward compatible. We could change it later.

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu (hunk 8 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu]: added logic to support it.

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 9]
```diff
@@ -126,30 +257,32 @@ Status QkvToContext(
     if (data.bias == nullptr) {
       // gemm_buffer should be BxSx3xNxH => qkv: 3xBxNxSxH
       ORT_ENFORCE(qk_head_size == v_head_size);
+      unsigned qkv_mask = (past_present_share_buffer ? 1 : 7);
       ORT_RETURN_IF_ERROR(LaunchTransQkv(stream, 3, sequence_length, batch_size, qk_head_size, num_heads,
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:261]: >LaunchTransQkv [](http://example.com/codeflow?start=26&length=14)

Do you need qkv_mask? How about to call LaunchTransQkv with num_of_matrix equal to 1 if past_present_share_buffer.

Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:261]: do not want to change too much on existing implementation. 3 is related with next stride to compute next position. There exists other ways with add more changes to this function, may worth to try if found useful.

**Review by yufenglee — DISMISSED** [review]
:shipit:

**Review by yufenglee — DISMISSED** [review]

**Review by yufenglee — COMMENTED** [review]
There is a build break on GPU and test failure on CPU

**Review by yufenglee — APPROVED** [review]
:shipit:

**Review by yufenglee — DISMISSED** [review]

**Review by hariharans29 — APPROVED** [review]

File: docs/ContribOperators.md [hunk 10]
```diff
@@ -143,13 +145,15 @@ This version of the operator has been available since version 1 of the 'com.micr
 <dt><tt>mask_index</tt> (optional) : M</dt>
 <dd>Attention mask with shape (batch_size, 1, max_sequence_length, max_sequence_length), (batch_size, total_sequence_length) or (batch_size, sequence_length, total_sequence_length), or index with shape (batch_size) or (2 * batch_size).</dd>
 <dt><tt>past</tt> (optional) : T</dt>
-<dd>past state for key and value with shape (2, batch_size, num_heads, past_sequence_length, head_size)</dd>
+<dd>past state for key and value with shape (2, batch_size, num_heads, past_sequence_length, head_size)When past_present_share_buffer is set, its shape is (2, batch_size, num_heads, max_sequence_length, head_size)</dd>
```
Comment (tianleiwu) [inline docs/ContribOperators.md:148]: Add some punctuation and space between two sentences.

File: docs/ContribOperators.md [hunk 11]
```diff
@@ -143,13 +145,15 @@ This version of the operator has been available since version 1 of the 'com.micr
...
 <dt><tt>extra_add</tt> (optional) : T</dt>
 <dd>additional add to QxK' with shape (batch_size, num_heads, sequence_length, total_sequence_length)</dd>
 <dt><tt>key</tt> (optional) : T</dt>
 <dd>Input for key with shape (batch_size, kv_sequence_length, hidden_size). Required when weights is not available.</dd>
 <dt><tt>value</tt> (optional) : T</dt>
 <dd>Input for key with shape (batch_size, kv_sequence_length, v_hidden_size). Required when weights is not available.</dd>
+<dt><tt>past_sequence_length</tt> (optional) : M</dt>
+<dd>When past_present_share_buffer, specify past_sequence_length for effective past sequence lenght (could be 0).Needed when past_present_share_buffer is not zero.</dd>
```
Comment (tianleiwu) [inline docs/ContribOperators.md:156]: lenght => length. Add a space before "Needed". (Need fix in code and re-generate the document by --gen_doc)

**Pull request merged**